### PR TITLE
resolve 'docker swarm leave -f' hang

### PR DESCRIPTION
**Pull request by ityangchen**
Signed-off-by: ityangchen <yangchenliang@huawei.com>

**- What I did**
please see , https://github.com/moby/moby/issues/35337

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks @ityangchen it looks like there's changes in this PR that need to be upstreamed first into the libnetwork and swarm kit repository. We may also want a test for this, but let me first ping maintainers for SwarmKit and Libnetwork

ping @anshulpundir @nishanttotla @fcrisciani

**Comment by thaJeztah** [comment]
ping @fcrisciani could you have a look at this one?

**Comment by thaJeztah** [comment]
Discussing this in the maintainers meeting, and there was some relation to the logic used in https://github.com/moby/moby/pull/32906 a well

File: vendor/github.com/docker/libnetwork/controller.go [hunk 1]
```diff
@@ -343,6 +343,7 @@ func (c *controller) clusterAgentInit() {
 			c.agentClose()
 			c.cleanupServiceBindings("")
 
+			c.agentInitComplete()
```
Comment (fcrisciani) [inline vendor/github.com/docker/libnetwork/controller.go]: not super convinced by this, in reality on the node leave if the agent did not complete the activation this should not be released anyway. Is there a particular reason for this here?

Comment (ityangchen) [reply vendor/github.com/docker/libnetwork/controller.go]: This is a temporary solution to resolve the `hang` issue only by testing.  I don`t know whether it is right, so seek community idea.

Comment (fcrisciani) [reply vendor/github.com/docker/libnetwork/controller.go]: if you remove it does it hangs still? if so can you send the SIGUSR1 and post the go routine stack trace?

Comment (ityangchen) [reply vendor/github.com/docker/libnetwork/controller.go]: Docker may hang when running for a long time if remove it . So sorry, the test environment be destroyed.

Comment (fcrisciani) [reply vendor/github.com/docker/libnetwork/controller.go]: if you run it for a long time like with no swarm init/swarm leave happening?

Comment (ityangchen) [reply vendor/github.com/docker/libnetwork/controller.go]: It can not be reproducted with no swarm init/swarm leave repeatedly.

**Comment by cpuguy83** [comment]
What's the status on this?

**Comment by olljanat** [comment]
@ityangchen this one needs to be vendoring

**Comment by cpuguy83** [comment]
Going to close this since it has gone stale.
If this change is still needed, please let us know and we can re-open.

Thanks!

**Pull request closed without merging**